src/Data/LLVM/BitCode/BitString.hs

```diff
@@ -33,35 +40,62 @@ pattern Bytes' n = NumBytes n
 bitCount :: NumBits -> Int
 bitCount (NumBits n) = n
 
+{-# INLINE addBitCounts #-}
 addBitCounts :: NumBits -> NumBits -> NumBits
 addBitCounts (NumBits a) (NumBits b) = NumBits $ a + b
 
+{-# INLINE subtractBitCounts #-}
 subtractBitCounts :: NumBits -> NumBits -> NumBits
 subtractBitCounts (NumBits a) (NumBits b) = NumBits $ a - b
 
+{-# INLINE bytesToBits #-}
+bitsToBytes :: NumBits -> (NumBytes, NumBits)
+bitsToBytes (NumBits n) = (NumBytes $ n `shiftR` 3, NumBits $ n .&. 7)
+
+{-# INLINE bitsToBytes #-}
+bytesToBits :: NumBytes -> NumBits
+bytesToBits (NumBytes n) = NumBits $ n `shiftL` 3
+
+----------------------------------------------------------------------
+
 data BitString = BitString
   { bsLength :: !NumBits
   , bsData   :: !Integer
-  } deriving Show
+  } deriving (Show, Eq)
+
+-- | Create an empty BitString
 
-instance Eq BitString where
-  BitString n i == BitString m j = n == m && i == j
+emptyBitString :: BitString
+emptyBitString = BitString (NumBits 0) 0
 
-instance Semigroup BitString where
-  BitString n@(NumBits n') i <> BitString m j =
-    BitString (addBitCounts n m) (i .|. (j `shiftL` n'))
 
-instance Monoid BitString where
-  mempty = BitString (NumBits 0) 0
-  mappend = (<>)
```

Comment:
I'm not sure how to feel about removing these instances, as they seem quite useful. Do later changes prevent these from being feasible?

Reply:
Yes, later changes make these irrelevant, and removing these instances helped to explicitly clarify the API boundary.  If these reasons weren't in play, I would agree that keeping these would be useful.